math/optimized_trig.h

```diff
@@ -0,0 +1,233 @@
+   bool is3rdOr4thQuadrant = angle_FP & FP_PI_Q;
+
+   // By checking whether there is a bit in the PI/2 place, we can determine if the angle is in the 2nd or 4th quadrant
+   bool is2ndOr4thQuadrant = angle_FP & FP_PI_HALVES_Q;
+
+   // Use the absolute value as cos(x) = cos(-x). Use a mask to truncate the value between 0 and pi.
+   int angle_truncated_FP = abs(angle_FP) & (FP_PI_Q-1);  // qQ
+
+   // Handle the special case of n*pi/2, which causes trouble for integer overflow
+   if (angle_truncated_FP == FP_PI_HALVES_Q) {
+      // Sine(+-pi/2) changes sign based on the quadrant
+      if (is3rdOr4thQuadrant == true) {
+         *sin_out = -1;
+      } else {
+         *sin_out = 1;
+      }
+
+      *cos_out = 0;
+      return;
+   }
+
+   // Use symmetry to calculate across [pi/2, pi). This is handy because the taylor coefficients are optimized for in this range.
+   const int32_t Q = angle_truncated_FP < FP_PI_HALVES_Q ? angle_truncated_FP : FP_PI_Q - angle_truncated_FP;  // qQ
+
+   // Shift X for proper scaling for use in the fixed-point math
+   int32_t X = Q >> (qQ-qX);  //qX
+
+   // Cosine(x). This MUST be done first
+   {
+      int32_t output;  // qN
+
+      // If the angle is close to pi/2, then use the simple approximation cos(pi/2-eps) ~= pi/2-eps
+      if (Q > (int32_t)(FP_PI_HALVES_Q*.95)) {
+         output = M_PI*(float)(((1<<(qQ-1))-Q) >> (-qN+qQ));  // qN
+      } else {
+         // --Calculate the 6th order approximation of cos(x)--//
+
+         int32_t X2 = X*X;                 // 2*qX
+         int32_t X2_16 = X2 >> (2*qX-16);  // 16
+
+         output = cosine6TaylorApproximation(AQ, BQm, CQ, DQm, X2, X2_16);
+      }
+
+      // Check if the original angle is in the 2nd or 3rd quadrant
+      if ((is2ndOr4thQuadrant && !is3rdOr4thQuadrant) || // the 2nd quadrant
+          (!is2ndOr4thQuadrant && is3rdOr4thQuadrant)) { // the 3rd quadrant
+         output = -output;
+      }
+
+      // Scale the output from qN to a float
+      *cos_out = (float)output / (1<<qN);
+   }
+
+   // Sine *must* be done after cosine. This is because the sine routine modifies `X`.
+   {
+      int32_t output = -1;
+
+      // If the angle is close to 0, then use the simple approximation sin(eps) ~= eps
+      if (Q < (int32_t)(FP_PI_HALVES_Q*.05)) {
+         output = M_PI*(float)(Q >> (-qN+qQ));  // qN
+      } else {
+         // --Calculate the 6th order approximation of sin(x)--//
+
+         // Subtract pi/2, exploiting sin(x) = cos(x-pi/2)
+         X -= FP_PI_HALVES;
+
+         int32_t X2 = X*X;                 // 2*qX
+         int32_t X2_16 = X2 >> (2*qX-16);  // 16
+
+         output = cosine6TaylorApproximation(AQ, BQm, CQ, DQm, X2, X2_16);
+      }
+
+      // Check if the original angle is in the 3rd or 4th quadrant
+      if (is3rdOr4thQuadrant == true) {
+         output = -output;
+      }
+
+      // Scale the output from qN to a float
+      *sin_out = (float)output / (1<<qN);
```

Comment:
(you can use multiplication with inverse here. gcc has to use division unless relaxed math is enables)

Reply:
I'll do that. The current output with -O2 is smart enough to recognize that this is a conversion to floating-point with a shift, but the multiplication solves any reliance on the compiler to do the right thing.

```
      *sin_out = (float)output / (1<<qN);
 8028ac2:	ee07 3a90 	vmov	s15, r3
 8028ac6:	eefa 7ac2 	vcvt.f32.s32	s15, s15, #28
 8028aca:	edc4 7a0e 	vstr	s15, [r4, #56]	; 0x38
```

Reply:
This looks reasonable as a solution:

```
      // Scale the output from qN to a float
#if __thumb2__  // Only compile this branch when the thumb instructions are present
      *sin_out = (float)output / (1<<qN);
#else
      const float den = 1/(1<<qN);
      *sin_out = output * den;
#endif
```